Image upload for KYC

Moejoe90 opened the pull request.

Moejoe90 committed:
Basic Upload

Moejoe90 commented on src/components/Upload.js:
```diff
@@ -0,0 +1,20 @@
+  children: PropTypes.node.isRequired,
+};
+
+export default Upload;
```
This is an abstract component that can be styled as we want by overriding "style"

recmo replied:
Ok. We should give it a style then that fits with the rest of the design. It should have a preview of the image, etc.

It also needs it wrapped in a Field from redux forms, so that we can include it in a form. The value of the component can be stored as a [DataURI](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs).

recmo commented on src/components/yarn.lock:
```diff
@@ -1,4 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
This file should never have existed. We missed it in this PR: https://github.com/Neufund/generic-ico/pull/27/files#diff-053e2c70f5644e6dbba259f10b013aa4

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,20 @@
+  (<Dropzone
+    {...props}
+    // What do we accept as valid?
+    accept="image/jpeg, image/png"
```
Does it support `image/*` or something similar?

Moejoe90 committed:
-Added preview for upload component
-Added Upload style sheet
-Designed upload component
-Modified upload component to redux-form

Moejoe90 commented on src/components/Upload.js:
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
```
Is there an example of how to connect i18 with redux state somewhere in the genric ico?

Moejoe90 commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+export default Upload;
+
+// TODO: add translation
+// TODO: add PDF support with special icon
```
Will there be pdf uploading?

recmo replied:
If we do, it will be a separate component. PDFs are not images.

Moejoe90 replied:
Ok i just got confused because in The UX Design design there was PDF as an accepted extension

recmo replied:
Ah, ok. The problem with PDFs is that they can not be used in an `<img src=` tag. So this is a situation where technical requirements are going to hurt the user experience, and we will have to make compromises.

@alxska We can either 1) reject pdf. 2) accept pdf but don't show preview or 3) spend two weeks building a pdf-rendering service that will show a preview after a few seconds delay.
Preference?

alxska replied:
@Recmo I already talked to Moe, my recommendation is either 1 or 2. 
In case of 2 I offered a solution to preview an icon with a file name instead of a picture.
https://material.io/icons/#ic_picture_as_pdf

Moejoe90 replied:
@remco Yes that would be the case if we choose to add .pdf then a pdf icon would be nominal and a conditional case would be added where its <img if its an image and <google icon lib if pdf
Now the question would a pdf format be acceptable as KYC or not, for example if in the future we chose to change to somehow an automatic validation system i would assume that adding pdf would complicate things.
I might be mistaken

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
```
We should name the component `FieldImage`.

Moejoe90 replied:
Done

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+        <p className={styles.header} >
+          <i className={`material-icons ${styles.logo}`}>publish</i> &nbsp; &nbsp; Click here to upload a file </p>
+        <p className={styles.body} >
+         Your file should not be bigger than 3 MB it can be .png .jpg .pdf
```
We should not accept PDF

Moejoe90 replied:
Done

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+    >
+      <div>
+        <p className={styles.header} >
+          <i className={`material-icons ${styles.logo}`}>publish</i> &nbsp; &nbsp; Click here to upload a file </p>
```
All user visible strings need to be translatable.

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+import PropTypes from 'prop-types';
+import styles from './Upload.scss';
+
+const renderDropzone = field => (
```
This is a component, so should start with a Capital letter. It should also have a descriptive name, like `DropzoneImage`.

Moejoe90 replied:
Done

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+import PropTypes from 'prop-types';
+import styles from './Upload.scss';
+
+const renderDropzone = field => (
```
You can use destructing on the `field` parameter, like `{name, input}`.

Moejoe90 replied:
The problem with that is then ES lint starts complaining about props validation and doing props validation for this component somehow feels like an over kill or?

recmo replied:
Validating input is an area where it's hard to do overkill, more checks are almost always better.

That ES lint doesn't complain now is a shotcomming in ES lint.

You can check the redux-form docs if they provide you we predefined proptypes.

recmo commented on src/components/Upload.js:
```diff
@@ -0,0 +1,46 @@
+      maxSize="3000"
+      className={styles.overide}
+      name={field.name}
+      onDrop={fileToUpload => field.input.onChange(fileToUpload)}
```
What is the format of the value produced by this field? If I use it in a form, what kind of data do I end up with?

Moejoe90 replied:
Field is actually the props sent from Field component 
Field will always send two props when dealing with components 
 [META,INPUT These are props that Field will pass to your wrapped component. The props provided by redux-form are divided into input and meta objects.](http://redux-form.com/6.0.5/docs/api/Field.md/)
In our case Input.value holds an array of 
[FILE](https://developer.mozilla.org/en/docs/Web/API/File) Objects. With File objects we can store the uploaded photos

recmo replied:
My questions was about the form field value. If I use this component in a form, what kind of value do a receive from this field in my `onSubmit` handler?

But we already discussed this. Currently it starts with an empty string, and then turns into a Javascript File object. This need to change to a dataUrl.

Moejoe90 committed:
-Component resturns base URL 64
-FILE handeling

Moejoe90 committed:
-Added Translations
-Component changes

z

Moejoe90 commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,64 @@
+import { getTranslator } from '../reducers/translation';
+import styles from './FieldImage.scss';
+
+const fileInstance = new FileReader(); // eslint-disable-line no-undef
```
[Support for this Library](https://caniuse.com/#search=Filereader) 
Is this a valid approach?

recmo replied:
Yes, it's the right WebAPI to use :+1:

Moejoe90 replied:
I was not sure about disabling ES-lint, i felt really bad about it

recmo replied:
A better way to solve it is by putting `/* global FileReader */` at the top before the imports.

Moejoe90 commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,64 @@
+    >
+      <div>
+        <p className={styles.header} >
+          <i className={`material-icons ${styles.logo}`}>{i18n('publish')}</i> &nbsp; &nbsp; {i18n('Click here to upload a file')} </p>
```
Text taken from UI

recmo replied:
Ok, but why the non-breaking spaces?

recmo replied:
`<i>` tag should not be used. Please read the notes in [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/i)

Moejoe90 replied:
Done

Moejoe90 commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,64 @@
+  i18n: PropTypes.func.isRequired,
+  input: PropTypes.shape({
+    name: PropTypes.string.isRequired,
+    value: PropTypes.oneOfType([
```
Currently it starts with an empty string, and then turns into a DataURL base64

recmo commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,62 @@
+      name={input.name}
+      onDrop={(imageData) => {
+        fileInstance.readAsDataURL(imageData[0]);
+        fileInstance.onloadend = () => input.onChange(fileInstance.result);
```
You are re-using the same global instance. What happens if we have multiple `<DropzoneImage>` components in a form?

Moejoe90 replied:
Do you mean fileInstance?
Currently if we have multiple `<FieldImage>` components in the same forum. They will work independently and store data with their given `<name>` attributes.

recmo commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,62 @@
+        <p className={styles.header} >
+          <i className={`material-icons ${styles.logo}`}>{i18n('publish')}</i> &nbsp; &nbsp; {i18n('Click here to upload a file')} </p>
+        <p className={styles.body} >
+          {i18n('Your file should not be bigger than 3 MB it can be')} .png .jpg
```
Why is `.png .jpg` not in the translated?

recmo commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,62 @@
+        { <img
+          style={{ maxHeight: '200px' }}
+          alt={input.name}
+          src={input.value}
```
What happens with PDF files here?

Moejoe90 replied:
In this iteration only images are accepted line 14 `accept={'image/*'}
I plan to add PDF as a separate pull request
`

Moejoe90 commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,64 @@
+import { getTranslator } from '../reducers/translation';
+import styles from './FieldImage.scss';
+
+const fileInstance = new FileReader(); // eslint-disable-line no-undef
```
I meant disabling ES-lint and using a global variable

recmo commented on src/components/FieldImage.js:
```diff
@@ -0,0 +1,57 @@
+      className={styles.overide}
+      name={input.name}
+      onDropAccepted={(imageData) => {
+        fileInstance.readAsDataURL(imageData[0]);
```
The same `fileInstance` is reused, this can cause a race condition.

Solution: A new instance needs to be created for every call to `onDropAccepted`.

Moejoe90 replied:
Interesting point, that would have bin a hidden issue.

Moejoe90 committed:
Fixed Global file instance creation
